src/gurt/misc.c

```diff
@@ -1192,3 +1197,145 @@ d_vec_pointers_append(struct d_vec_pointers *pointers, void *pointer)
 	pointers->p_len++;
 	return 0;
 }
+
+/**
+ * Overloads to hook the unsafe getenv()/[un]setenv()/putenv()/clearenv()
+ * functions from glibc.
+ * Libgurt is the preferred place for this as it is the lowest layer in DAOS,
+ * so it will be the earliest to be loaded and will ensure the hook to be
+ * installed as early as possible and could prevent usage of LD_PRELOAD.
+ * The idea is to strengthen all the environment APIs by using a common lock.
+ *
+ * XXX this will address the main lack of multi-thread protection in the Glibc
+ * APIs but do not handle all unsafe use-cases (like the change/removal of an
+ * env var when its value address has already been grabbed by a previous
+ * getenv(), ...).
+ */
+
+static pthread_mutex_t hook_env_lock = PTHREAD_MUTEX_INITIALIZER;
+static char *(*real_getenv)(const char *);
+static int (*real_putenv)(char *);
+static int (*real_setenv)(const char *, const char *, int);
+static int (*real_unsetenv)(const char *);
+static int (*real_clearenv)(void);
+
+char *getenv(const char *name)
+{
+	char *p;
+
+	D_MUTEX_LOCK(&hook_env_lock);
+	if (real_getenv == NULL) {
+		real_getenv = (char * (*)(const char *))dlsym(RTLD_NEXT, "getenv");
+		if (real_getenv == NULL) {
+			/* Glibc symbols could not be resolved !!... */
+			void *handle;
+
+			handle = dlopen("libc.so.6", RTLD_LAZY);
+			D_ASSERT(handle != NULL);
+			real_getenv = (char * (*)(const char *))dlsym(handle, "getenv");
+		}
+		D_ASSERT(real_getenv != NULL);
+	}
+
+	p = real_getenv(name);
```

Comment:
> If you separate init from this, the lock here should really only protect the real call.

Well, initialisation of the "real_..." variables is racy too, but may be I can switch to using "atomic" variables then ?!

Reply:
Finally, I have switched to using Atomics for the initialisation part and also switched to rw-lock usage (instead of simple mutex, mainly to allow concurrent getenv()s) for the exception part.